staking/economics/reward.go

```diff
@@ -0,0 +1,284 @@
+		if err != nil {
+			return nil, nil, err
+		}
+		stakedNow = stakedNow.Add(
+			numeric.NewDecFromBigInt(wrapper.TotalDelegation()),
+		)
+	}
+	percentage := stakedNow.Quo(totalTokens.Mul(
+		reward.PercentageForTimeStamp(timestamp),
+	).Add(dole))
+	utils.Logger().Info().
+		Str("so-far-doled-out", dole.String()).
+		Str("staked-percentage", percentage.String()).
+		Str("currently-staked", stakedNow.String()).
+		Msg("Computed how much staked right now")
+	return soFarDoledOut.NetworkTotalPayout, &percentage, nil
+}
+
+// newSnapshot returns a record with metrics on
+// the network accumulated rewards,
+// and by validator.
+func newSnapshot(
+	beaconchain engine.ChainReader, header *block.Header, includeAPRs bool,
+) (*Snapshot, error) {
+	blockNow := header.Number().Uint64()
+	blocksPerEpoch := shard.Schedule.BlocksPerEpoch()
+	oneEpochAgo := beaconchain.GetHeaderByNumber(blockNow - blocksPerEpoch)
+	twoEpochAgo := beaconchain.GetHeaderByNumber(blockNow - (blocksPerEpoch * 2))
+
+	if !beaconchain.Config().IsStaking(header.Epoch()) {
+		return nil, errors.Errorf(
+			"current epoch %s, staking epoch %s",
+			header.Epoch().String(),
+			beaconchain.Config().StakingEpoch.String(),
+		)
+	}
+
+	soFarDoledOut, err := beaconchain.ReadBlockRewardAccumulator(blockNow)
+
+	if err != nil {
+		return nil, err
+	}
+
+	stakedNow, rates, junk :=
+		numeric.ZeroDec(), []ComputedAPR{}, numeric.ZeroDec()
+	// Only active validators' stake is counted in
+	// stake ratio because only their stake is under slashing risk
+	active, err := beaconchain.ReadActiveValidatorList()
+	if err != nil {
+		return nil, err
+	}
+
+	if includeAPRs {
+		rates = make([]ComputedAPR, len(active))
+	}
+
+	dole := numeric.NewDecFromBigInt(soFarDoledOut.NetworkTotalPayout)
+
+	for i := range active {
+		wrapper, err := beaconchain.ReadValidatorInformation(active[i])
+		if err != nil {
+			return nil, err
+		}
+		total := wrapper.TotalDelegation()
+		stakedNow = stakedNow.Add(numeric.NewDecFromBigInt(total))
+		if includeAPRs {
+			rates[i] = ComputedAPR{
+				Validator:        active[i],
+				TotalStakedToken: total,
+				// these two last fields meant to be mutated later.
+				StakeRatio: junk,
+				APR:        numeric.ZeroDec(),
+			}
+		}
+	}
+
+	circulatingSupply := totalTokens.Mul(
+		reward.PercentageForTimeStamp(header.Time().Int64()),
+	).Add(dole)
+
+	if oneEpochAgo != nil && twoEpochAgo != nil {
+		if blocksPerYear, err := ExpectedValueBlocksPerYear(
+			oneEpochAgo, twoEpochAgo, int64(blocksPerEpoch),
+		); includeAPRs && err != nil {
+			rewardsSoFar := soFarDoledOut.ValidatorRewards
+			soFarCount := len(rewardsSoFar)
+
+			sort.SliceStable(rewardsSoFar, func(i, j int) bool {
+				return bytes.Compare(
+					rewardsSoFar[i].Validator.Bytes(),
+					rewardsSoFar[j].Validator.Bytes(),
+				) == -1
+			})
+
+			for i := range rates {
+				lookingFor := rates[i].Validator.Bytes()
+				if k :=
+					sort.Search(soFarCount, func(j int) bool {
```

Comment:
This part is too complicated too. Basically validator's APR is estimated_reward_per_year/total_stake; estimated_reward_per_year=avg_reward_per_block * blocks_per_year; avg_reward_per_block = total_reward_per_epoch / blocks_per_epoch;